library/platform_util.c

```diff
@@ -172,51 +183,79 @@ int mbedtls_platform_memcmp( const void *buf1, const void *buf2, size_t num )
     return( (int) diff | (int) ( flow_counter ^ num ) );
 }
 
-uint32_t mbedtls_platform_random_uint32( )
+/* This function implements a non-cryptographic random number generator based
+ * on the linear congruential generator algorithm. Additionally, if the
+ * MBEDTLS_ENTROPY_HARDWARE_ALT flag is defined, the seed is changed with using
+ * a hardware random number generator every RAND_SEED_LIFE number of iterations.
+ *
+ * The value of the returned number is in the range [0; 0xffff].
+ */
+static uint32_t mbedtls_platform_random_uint16( void )
 {
-#if !defined(MBEDTLS_ENTROPY_HARDWARE_ALT)
-    return 0;
-#else
-    uint32_t result = 0;
-    size_t olen = 0;
+    /* Set random_state - the first random value should not be zero. */
+    static uint32_t random_state = RAND_INCREMENT;
 
-    mbedtls_hardware_poll( NULL, (unsigned char *) &result, sizeof( result ),
-                           &olen );
-    return( result );
-#endif
+#if defined(MBEDTLS_ENTROPY_HARDWARE_ALT)
+
+    static uint32_t random_seed_life = 0;
+
+    if( 0 < random_seed_life )
+    {
+        --random_seed_life;
+    }
+    else
+    {
+        size_t olen = 0;
+        uint32_t hw_random;
+        mbedtls_hardware_poll( NULL,
+                               (unsigned char *) &hw_random, sizeof( hw_random ),
+                               &olen );
+        if( olen == sizeof( hw_random ) )
+        {
+            random_state ^= hw_random;
+            random_seed_life = RAND_SEED_LIFE;
+        }
+    }
+
+#endif /* MBEDTLS_ENTROPY_HARDWARE_ALT */
+
+    random_state = ( ( random_state * RAND_MULTIPLIER ) + RAND_INCREMENT ) % RAND_MODULUS;
+
+    /* Do not return the entire random_state to hide generator predictability for
+     * the next iteration */
+    return( ( random_state >> 15 ) & 0xffff );
 }
 
-uint32_t mbedtls_platform_random_in_range( size_t num )
+uint32_t mbedtls_platform_random_uint32( void )
 {
-#if !defined(MBEDTLS_ENTROPY_HARDWARE_ALT)
-    (void) num;
-    return 0;
-#else
-    uint32_t result = 0;
-    size_t olen = 0;
+    return( ( mbedtls_platform_random_uint16() << 16 ) |
+              mbedtls_platform_random_uint16() );
+}
 
-    mbedtls_hardware_poll( NULL, (unsigned char *) &result, sizeof( result ),
-                           &olen );
+uint32_t mbedtls_platform_random_in_range( uint32_t num )
+{
+    uint32_t result;
 
-    if( num == 0 )
+    if( num <= 1 )
     {
         result = 0;
     }
+    else if( num <= 0xffff )
+    {
+        result = mbedtls_platform_random_uint16() % num;
```

Comment:
@piotr-now ת why do we need mbedtls_platform_random_uint16 ? Why we cannot use always mbedtls_platform_random_uint32 ?

Reply:
Actually, we don't need mbedtls_platform_random_uint16(). This is only related to optimization because mbedtls_platform_random_in_range() is most often called for values below 0xffff and one call of mbedtls_platform_random_uint16() is enough. For values above 0xffff the mbedtls_platform_random_uint16() function must be called twice and this is done by mbedtls_platform_random_uint32()

Reply:
But you always generate uint32_t in hw poll function.  What am i missing?

Reply:
Anyway, lets generate always 32. It will save as a little bit code size

Reply:
Done

Reply:
@piotr-now , But there is still random32 functions that calls to random16 twice. And random16 function returns 32bit result twice.  Do we have security reason for double random ?

Reply:
There are two reasons why I did this. The first results from the modulo value in the random generator. If the generator were to return the uint32_t value, the modulo value would have to be 2^32 (for the 32-bit arch the modulo value would not be at all). The second reason was more important to me and it is that the random generator would have to return the entire value of the `random_state` variable. It is not a very good practice because knowing the generator parameters and the last returned random value, you can easily calculate the entire sequence of next generated values.

Reply:
@piotr-now . I have read your and Alexander's explanations.Thank you :)
 Let's only add comment to explain

Reply:
No problem :). I have added an explanation about this 16-bit value.